chinglee-iot/FreeRTOS-Kernel
# Add support for portCRITICAL_NESTING_IN_TCB disabled

Description
-----------
This POC removes the portCRITICAL_NESTING_IN_TCB constraint in SMP.

The problems are the following variables in kernel are private:
1. critical nesting count : prvCheckForRunStateChanged needs this variable to decide if the task just entering the critical section or suspending the scheduler
2. xTaskRunState in TCB : prvCheckForRunStateChanged needs this variable to know if this task is requested to yield when it disable the interrupt and trying to acquire the lock
3. xYieldPendings : Task yields when leaving the critical section

This PR separates the behavior into porting and kernel.
The code need to access kernel variables is left in kernel. The porting only maintains it's own critical nesting count.

Porting export the macros to kernel 
```
#define portGET_CRITICAL_NESTING_COUNT()          ( uxCriticalNestings[ portGET_CORE_ID() ] )
```

Two functions are added for SMP in order to support critical nesting count in porting.
* vSmpTaskEnterCritical() : Task need to check run state change when just entering the critical section. If the run state changed to yield, the task need to exit the critical section and try to enter the critical section again.
* vSmpTaskExitCritical() : Task need to check xYieldPendings when leaving the critical section. If there is pending yield, the task need to yield immediately.

Test Steps
-----------
Run RP2040 demo with portCRITICAL_NESTING_IN_TCB = 0

Related Issue
-----------



By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
